addons/binding/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/handler/AbstractOmnilinkHandler.java

```diff
@@ -1,26 +1,88 @@
 package org.openhab.binding.omnilink.handler;
 
+import java.util.Optional;
+
 import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
 import org.eclipse.smarthome.core.thing.ThingUID;
 import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
 import org.openhab.binding.omnilink.OmnilinkBindingConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.digitaldan.jomnilinkII.MessageTypes.statuses.Status;
 
 /**
  *
  * @author Craig Hamilton
  *
  */
-public abstract class AbstractOmnilinkHandler extends BaseThingHandler {
+public abstract class AbstractOmnilinkHandler<T extends Status> extends BaseThingHandler {
+
+    private static Logger logger = LoggerFactory.getLogger(AbstractOmnilinkHandler.class);
 
     public AbstractOmnilinkHandler(Thing thing) {
         super(thing);
     }
 
+    private volatile Optional<T> status;
+
     public OmnilinkBridgeHandler getOmnilinkBridgeHander() {
         return (OmnilinkBridgeHandler) getBridge().getHandler();
     }
 
+    @Override
+    public void initialize() {
+        Optional<T> status = retrieveStatus();
+        handleStatus(status.orElse(null)); // handle status will process null.
+        super.initialize();
+    }
+
+    /**
+     * Attempt to retrieve an updated status for this handler type.
+     *
+     * @return Optional with updated status if possible, empty optional otherwise.
+     */
+    protected abstract Optional<T> retrieveStatus();
+
+    /**
+     * Update channels associated with handler
+     *
+     * @param t Status object to update channels with
+     */
+    protected abstract void updateChannels(T t);
+
+    /**
+     * Process a status update for this handler. This will dispatch updateChannels where appropriate.
+     *
+     * @param t Status to process.
+     */
+    public void handleStatus(T t) {
+        this.status = Optional.ofNullable(t);
+        if (status.isPresent()) {
+            updateChannels(status.get());
```

Comment:
@boc-tothefuture, I'm trying to figure out if the updateChannels call should have no argument, and instead just access the status instance variable.  As I see it, if handleStatus is called with null, then the status instance variable will be switched to Empty, but the channels will still reflect the prior status if there was one.  Do you see pros/cons vs arg/no arg?

Reply:
This will come down to personal preference.  In general, I prefer not to leak fields to subclasses.  I also tend to be very contract oriented in the method calls for subclasses.  What I mean by that is, if I call a protected abstract method, I like to supply the argument for it to work with so it is clear to the subclass what I expect to happen - I think supplying that argument makes it clear what the channels should be updated with.  If we leak the status object to the subclasses, they will have to do all of their own checks to validate it can be processed (null/empty, etc), rather than do it in a single place.  Once again, personal opinion, so I am open to discussion.

So, you are correct, if the status is null, we now persist an empty optional and don't update the channels.  In the current implementations if the status is null, I think just throw an NPE.  I get your point that we have no received a state that doesn't match.  However, if we do get a null state, I don't think we would ever take any action (clear the channels, or something?).  So, perhaps the best approach is for the abstract handler to just log and ignore null status updates?  I am happy to make that change, let me know what you think.

Reply:
your position here makes sense.  definitely wouldn't want to make changes just based on style.  I like hearing how others handle inheritance etc as it can be a part of OO which really gets messy and complicated, so thanks for sharing.